### PR TITLE
FEAT-#6492: Add `from_map` feature to create dataframe

### LINKED ISSUE #6492: Equivalent of dask.dataframe.from_map?
dask.dataframe has an experimental function called [`from_map`](https://docs.dask.org/en/latest/generated/dask.dataframe.from_map.html). This allows the user to define a custom function for loading/creating partitions of a dataframe. Conceptually, I think this is similar to [Registering Custom Functions](https://github.com/modin-project/modin/blob/cf335c30758244309db9fc9e3f75c8f19618146f/examples/tutorial/jupyter/execution/pandas_on_ray/local/exercise_3.ipynb), except that it allows a function to **create** a DataFrame, rather than operate on an **existing** DataFrame.

This seems like it would be very useful to have in Modin; I don't see any straightforward way to achieve this currently. Am I missing something? I know that it would need to have some additional complexity that Dask doesn't have to worry about, in order to be general across storage formats and allow 2D partitioning, but these don't seem like insurmountable issues.

Any reactions to the utility and/or feasibility of this feature?

### PR DESCRIPTION
**Pull request by YarShev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6492  
- [x] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/execution/unidist/implementations/pandas_on_unidist/io/io.py [hunk 1]
```diff
@@ -258,3 +261,66 @@ def func(df, **kw):  # pragma: no cover
         UnidistWrapper.materialize(
             [part.list_of_blocks[0] for row in result for part in row]
         )
+
+    @classmethod
+    def from_map(cls, func, iterable, *args, **kwargs):
```
Comment (anmyachev) [inline modin/core/execution/unidist/implementations/pandas_on_unidist/io/io.py:266]: Is it possible to use already implemented functions with `num_splits=1`?

File: modin/pandas/io.py [hunk 2]
```diff
@@ -1109,6 +1109,36 @@ def from_dask(dask_obj) -> DataFrame:
     return ModinObjects.DataFrame(query_compiler=FactoryDispatcher.from_dask(dask_obj))
 
 
+def from_map(func, iterable, *args, **kwargs) -> DataFrame:
```
Comment (anmyachev) [inline modin/pandas/io.py:1112]: Documentation needs to be updated I suppose.

File: modin/core/execution/unidist/implementations/pandas_on_unidist/io/io.py (hunk 1 above)
Comment (YarShev) [reply modin/core/execution/unidist/implementations/pandas_on_unidist/io/io.py:266]: I don't quite get what would you like use use instead. Please elaborate. We are adding a new `from_map` by analogy with other io functions.

Comment (YarShev) [reply modin/core/execution/unidist/implementations/pandas_on_unidist/io/io.py:266]: I suppose we can't use anything from existing functionality as every method of a Modin Dataframe assumes there is a dataframe with partitions to apply a function to.

File: modin/pandas/io.py (hunk 2 above)
Comment (YarShev) [reply modin/pandas/io.py:1112]: We don't have docs for such methods as from_pandas, from_ray, from_dask, etc. Do you think we should update docs on this matter in a separate issue in one go?

Comment (anmyachev) [reply modin/pandas/io.py:1112]: ок

File: modin/tests/pandas/test_io.py [hunk 3]
```diff
@@ -3461,3 +3461,20 @@ def test_from_dask():
 
     result_df = from_dask(dask_df)
     df_equals(result_df, modin_df)
+
+
+@pytest.mark.skipif(
+    condition=Engine.get() not in ("Ray", "Dask", "Unidist"),
+    reason="Dask DataFrame can only be created if Modin uses Ray, Dask or MPI engine.",
```
Comment (anmyachev) [inline modin/tests/pandas/test_io.py:3468]: Wrong reason, right?

Comment (YarShev) [reply modin/tests/pandas/test_io.py:3468]:
```suggestion
    reason="Modin DataFrame can only be created from map if Modin uses Ray, Dask or MPI engine.",
```
Good catch, thanks!

File: modin/core/execution/ray/implementations/pandas_on_ray/io/io.py [hunk 4]
```diff
@@ -302,3 +305,66 @@ def to_ray(cls, modin_obj):
...
+        -------
+        BaseQueryCompiler
+            QueryCompiler containing data returned by map function.
+        """
+        func = cls.frame_cls._partition_mgr_cls.preprocess_func(func)
+        partitions = np.array(
+            [
+                [
+                    cls.frame_partition_cls(
+                        deploy_map_func.remote(func, obj, *args, **kwargs)
```
Comment (anmyachev) [inline modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:338]: I suggest to use `RayWrraper.deploy` here.

Comment (anmyachev) [reply modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:338]: And corresponding wrappers for other engines.

Comment (anmyachev) [reply modin/core/execution/ray/implementations/pandas_on_ray/io/io.py:338]: To reduce the likelihood of error, we need to either have all launch options in one place, or use only one method. There is a tendency that launching functions becomes more difficult due to additional parameters. A good example is `resources=RayTaskCustomResources.get()`, which is currently not taken into account here.

We can move this function to `engine_wrapper.py` and call it inside `Raywrapper.deploy` using an additional parameter.

Comment (YarShev) [reply modin/core/execution/ray/implementations/pandas_on_ray/io/io.py]: Re-used *.deploy.

File: modin/core/execution/dask/implementations/pandas_on_dask/io/io.py [hunk 5]
```diff
@@ -188,3 +190,45 @@ def df_to_series(df):
...
+                            func,
+                            f_args=(obj,) + args,
+                            f_kwargs=kwargs,
+                            return_pandas_df=True,
+                        )
+                    )
+                ]
+                for obj in iterable
+            ]
+        )
```
Comment (anmyachev) [inline modin/core/execution/dask/implementations/pandas_on_dask/io/io.py:233]: Based on the information required to perform this task, it seems that a more appropriate level at which to define the function would be a partition manager, for example somewhere around:

https://github.com/modin-project/modin/blob/c48bb30480f139f08d720171c6107dc74183edb6/modin/core/dataframe/pandas/partitioning/partition_manager.py#L186

File: modin/tests/pandas/test_io.py [hunk 6]
```diff
@@ -3461,3 +3461,20 @@ def test_from_dask():
 
     result_df = from_dask(dask_df)
     df_equals(result_df, modin_df)
+
+
+@pytest.mark.skipif(
+    condition=Engine.get() not in ("Ray", "Dask", "Unidist"),
```
Comment (anmyachev) [inline modin/tests/pandas/test_io.py:3467]: Would it be more correct to limit it not by engines, but by storage format: `pandas`?

**Review by anmyachev — APPROVED** [review]

File: modin/core/execution/dask/implementations/pandas_on_dask/io/io.py (hunk 5 above)
Comment (YarShev) [reply modin/core/execution/dask/implementations/pandas_on_dask/io/io.py:233]: I would leave it here. Imagine a case when iterable is a list files.

File: modin/tests/pandas/test_io.py (hunk 6 above)
Comment (YarShev) [reply modin/tests/pandas/test_io.py:3467]: PandasOnPython wouldn't work. Let's leave as is.

File: modin/core/execution/dask/implementations/pandas_on_dask/io/io.py (hunk 5 above)
Comment (anmyachev) [reply modin/core/execution/dask/implementations/pandas_on_dask/io/io.py:233]: > Imagine a case when iterable is a list files.

We'll be abstracting from the parameters just like we're doing now, so I don't see any difference.

File: modin/tests/pandas/test_io.py (hunk 6 above)
Comment (anmyachev) [reply modin/tests/pandas/test_io.py:3467]: > PandasOnPython wouldn't work.

As far as I can see, there are no restrictions on its operation. We just need to add essentially the same code as for the other engines.

File: modin/pandas/io.py (hunk 2 above)
Comment (anmyachev) [reply modin/pandas/io.py:1112]: @YarShev are you going to do this before release?

Comment (YarShev) [reply modin/pandas/io.py:1112]: Yes, that would be great - https://github.com/modin-project/modin/pull/7256.

**Pull request merged**